File: PathoFroh/src/main/java/com/patho/main/service/ListItemService.java
package com.patho.main.service;

import com.patho.main.model.ListItem;
import com.patho.main.model.ListItem_;
import com.patho.main.model.interfaces.ListOrder;
import com.patho.main.repository.ListItemRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Predicate;
import javax.persistence.criteria.Root;
import java.util.List;

@Service
@Transactional
public class ListItemService extends AbstractService {

	@Autowired
	private ListItemRepository listItemRepository;

	public ListItem addOrUpdate(ListItem listItem) {
		ListItem.StaticList type = listItem.getListType();
		if (listItem.getId() == 0) {

			long count = listItemRepository.count(new Specification<ListItem>() {
				@Override
				public Predicate toPredicate(Root<ListItem> root, CriteriaQuery<?> query, CriteriaBuilder cb) {
					return cb.equal(root.get(ListItem_.listType), type);
				}
			});

			System.out.println("lenght " + count);

			listItem.setIndexInList((int) count);

			logger.debug("Creating new ListItem " + listItem.getValue() + " for " + listItem.getListType().toString());

			listItem = listItemRepository.save(listItem, resourceBundle.get("log.settings.staticList.new",
					listItem.getValue(), listItem.getListType().toString()));

		} else {
			logger.debug("Updating ListItem " + listItem.getValue());
			listItem = listItemRepository.save(listItem, resourceBundle.get("log.settings.staticList.update",
					listItem.getValue(), listItem.getListType().toString()));
		}

		return listItem;
	}

	/**
	 * Tries to delete the old listItem
	 * 
	 * @param p
	 */
	@Transactional(propagation = Propagation.NEVER)
	public boolean deleteOrArchive(ListItem listItem) {
		try {
			listItemRepository.delete(listItem,
					resourceBundle.get("log.settings.staticList.deleted", listItem.getValue(), listItem.getListType()));
			return true;
		} catch (Exception e) {
			archive(listItem, true);
			return false;
		}
	}

	/**
	 * Archived or dearchives a ListItem
	 * 
	 * @param p
	 * @param archive
	 * @return
	 */
	public ListItem archive(ListItem listItem, boolean archive) {
		listItem.setArchived(archive);
		return listItemRepository.save(listItem,
				resourceBundle.get(archive ? "log.settings.staticList.archived" : "log.settings.staticList.dearchived",
						listItem.getValue(), listItem.getListType()));
	}

	public List<ListItem> updateReoderedList(List<ListItem> staticListContent, ListItem.StaticList type) {
		ListOrder.reOrderList(staticListContent);
		return listItemRepository.saveAll(staticListContent,
				resourceBundle.get("log.settings.staticList.list.reoder", type));
	}
}

File: PathoFroh/src/main/kotlin/com/patho/main/service/ListItemService.kt
package com.patho.main.service

import com.patho.main.model.interfaces.ListOrder
import com.patho.main.model.preset.ListItem
import com.patho.main.model.preset.ListItemType
import com.patho.main.model.preset.ListItem_
import com.patho.main.repository.jpa.ListItemRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional


@Service()
open class ListItemService @Autowired constructor(
        private val listItemRepository: ListItemRepository) : AbstractService() {

    /**
     * Adds or updates a list Item. If added the list item indexInList is set to last in list
     */
    @Transactional
    open fun addOrUpdate(listItem: ListItem): ListItem {
        var result = listItem
        if (result.id == 0L) {
            logger.debug("Creating new ListItem " + result.value + " for " + result.listType.toString())
            result.indexInList = listItemRepository.count { root, _, cb -> cb.equal(root.get(ListItem_.listType), result.listType) }.toInt()
            result = listItemRepository.save(result, resourceBundle["log.settings.staticList.new", result.value, result.listType.toString()])
        } else {
            logger.debug("Updating ListItem " + result.value)
            result = listItemRepository.save(result, resourceBundle["log.settings.staticList.update", result.value, result.listType.toString()])
        }
        return result
    }

    /**
     * Deletes or archives a ListItem. If deleting is not possible the item will be archvied.
     */
    @Transactional(propagation = Propagation.NEVER)
    open fun deleteOrArchive(listItem: ListItem): Boolean {
        return try {
            listItemRepository.delete(listItem, resourceBundle["log.settings.staticList.deleted", listItem.value, listItem.listType])
            true
        } catch (e: Exception) {
            archive(listItem, true)
            false
        }
    }

    /**
     * Archives a ListItem
     */
    @Transactional
    open fun archive(listItem: ListItem, archive: Boolean): ListItem {
        listItem.archived = archive
        return listItemRepository.save(listItem,
                resourceBundle[if (archive) "log.settings.staticList.archived" else "log.settings.staticList.dearchived", listItem.value, listItem.listType])
    }

    /**
     * Reorganizes and persists the list
     */
    @Transactional
    open fun updateReorderedList(staticListContent: List<ListItem>, type: ListItemType): List<ListItem> {
        ListOrder.reOrderList(staticListContent)
        return listItemRepository.saveAll(staticListContent,
                resourceBundle["log.settings.staticList.list.reoder", type])
    }
}